enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,491 @@
+---
+title: set-replace-append-delete-deny-headers
+authors:
+  - "@miheer"
+reviewers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+approvers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+api-approvers:
+  - "@JoelSpeed"
+  - "@deads2k"
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+creation-date: 2022-12-02
+last-updated: 2022-12-14
+tracking-link:
+  - "https://issues.redhat.com/browse/NE-982"
+  - "https://issues.redhat.com/browse/RFE-464"
+status: TODO
+see-also:
+  - ""
+replaces:
+superseded-by:
+---
+
+# Set, Delete, Deny HTTP Response headers
+
+## Summary
+
+This enhancement extends the IngressController API and Route API to allow the user to set or delete
+HTTP response headers. Using the new API, users may specify a list of
+HTTP response headers either via IngressController API or Route API that should be set or deleted.
+Also, this enhancement creates an immutable(hard-coded) list of HTTP response headers which can't be set or deleted.
+
+## Motivation
+
+Certain HTTP response headers may need to be set, deleted for compliance reasons or other reasons. In OpenShift 3, cluster 
+administrators could use a custom HAProxy configuration template to set or delete headers, but using a custom
+template is not permitted in OpenShift 4.
+Certain HTTP response headers like HSTS (Strict-Transport-Security) need to be denied as we already provide a way to set them
+via route annotation.
+
+### User Stories
+
+1. As a cluster administrator, I need to configure an IngressController to set/delete certain HTTP response headers in order to
+   satisfy compliance requirements at global level to send/not to send them to the client which has called any route from IngressController's domain.
+2. As a user, I need to configure Route to set/delete HTTP response headers to send/not send them to the client which has called the route.
+
+### Goals
+
+1. Enable the cluster administrator to specify HTTP response headers in the IngressController CR spec that should be set or 
+   deleted at the global level, i.e. for all routes.
+2. Enable the user to specify HTTP response headers in the Route spec (when it is not a passthrough Route) for their application
+   that should be set or deleted when forwarding a response to the client.
+3. Define an immutable/hard-coded deny list of headers which are not allowed to be set or deleted.
+   Eg- we should prohibit specifying strict-transport-security using the new API, for the following reasons:
+   HSTS is already configurable through a route annotation.
+   We already have an API for restricting HSTS in the cluster config.
+   Adding a second way to configure HSTS would significantly complicate this enhancement.
```

Comment:
Is it not just setting a particular header key and value?

Reply:
HSTS can be set via route annotation. However, the policy as to what can be set on a route annotation can be defined on the ingress config object.

Reply:
> Is it not just setting a particular header key and value?

The `strict-transport-security` header (AKA HSTS) has wider ranging implications than the typical HTTP header.  We considered allowing HSTS to be configured on the IngressController in the past; see [the global-options-enable-hsts EP](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/global-options-enable-hsts.md), which was never implemented.  However, HSTS isn't neatly scoped:  Route A's HSTS policy can affect Route B, even if the routes are in different namespaces or belong to different IngressControllers, if Route B's host name is a subdomain of Route A's.  Additionally, specifying HSTS on an IngressController can cause a route manifest that behaves as expected on one cluster to behave surprisingly and problematically on another cluster.  For these reasons, we decided not to implement the global-options-enable-hsts EP and instead wrote and implemented [the global-admission-hsts EP](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/global-admission-hsts.md), which defines a cluster-scoped API, using which the cluster admin can specify HSTS parameters that may or must be specified on routes for a given domain: <https://github.com/openshift/api/blob/6ba31fa438f20d6c822ef47ad3d771309cb5216b/config/v1/types_ingress.go#L64-L87>.  Based on this experience, we wouldn't want to allow configuring HSTS on the IngressController using this new API.

Reply:
Thanks Miciah, that's a really good justification for not allowing this, can we make sure this enhancement is updated to cover this so that future readers have that same context and don't ask the same question again

Reply:
done